In [1]:
# Setup()
# Compare_models(): Compare multiples modelos y los ranker 
# Create_model () : Crear un modelo específico
# Tune_models( ):  Ajuste automático del modelo
# Plot_model() : Visualizar el desempeño del modelo
# Evaluate_model() : Evaluación del modelo
# Predict_model() : Predicciones sobre nuevos datos en el Df.
# Save_model/load_model() : 

In [2]:
import pandas as pd 
import numpy as np 
import matplotlib.pyplot as plt
import seaborn as sns 

from pycaret.regression import *

In [3]:
df = pd.read_csv(r'C:\Users\Jeffersson Pretell\Documents\Especializacion_ML_Engineering\data\precio_departamentos.csv')

In [4]:
df

,area_m2,num_habitaciones,num_banos,distrito,antiguedad_anos,vista_mar,precio
0,139,1,2,San Isidro,7,0,380619.28
1,96,1,1,San Isidro,7,0,221302.53
2,122,4,1,San Isidro,23,0,338702.68
3,128,2,1,Surco,21,1,341815.62
4,132,3,2,Surco,19,0,346460.60
...,...,...,...,...,...,...,...
95,31,3,2,Miraflores,10,0,108020.65
96,81,2,1,Magdalena,27,0,200124.07
97,74,3,2,San Isidro,14,0,214778.48
98,147,3,2,Magdalena,15,0,417997.49


In [20]:
help(setup)

Help on function setup in module pycaret.regression.functional:

setup(data: Union[dict, list, tuple, numpy.ndarray, scipy.sparse._matrix.spmatrix, pandas.core.frame.DataFrame, NoneType] = None, data_func: Optional[Callable[[], Union[dict, list, tuple, numpy.ndarray, scipy.sparse._matrix.spmatrix, pandas.core.frame.DataFrame]]] = None, target: Union[int, str, list, tuple, numpy.ndarray, pandas.core.series.Series] = -1, index: Union[bool, int, str, list, tuple, numpy.ndarray, pandas.core.series.Series] = True, train_size: float = 0.7, test_data: Union[dict, list, tuple, numpy.ndarray, scipy.sparse._matrix.spmatrix, pandas.core.frame.DataFrame, NoneType] = None, ordinal_features: Optional[Dict[str, list]] = None, numeric_features: Optional[List[str]] = None, categorical_features: Optional[List[str]] = None, date_features: Optional[List[str]] = None, text_features: Optional[List[str]] = None, ignore_features: Optional[List[str]] = None, keep_features: Optional[List[str]] = None, preproces

In [14]:
df.nunique()

area_m2              75
num_habitaciones      4
num_banos             2
distrito              4
antiguedad_anos      28
vista_mar             2
precio              100
dtype: int64

In [5]:
# 1.- Setup()
reg = setup(data = df,target = 'precio',session_id = 123,normalize = True,categorical_features = ['distrito'])

,Description,Value
0,Session id,123
1,Target,precio
2,Target type,Regression
3,Original data shape,"(100, 7)"
4,Transformed data shape,"(100, 10)"
5,Transformed train set shape,"(70, 10)"
6,Transformed test set shape,"(30, 10)"
7,Numeric features,5
8,Categorical features,1
9,Preprocess,True


In [6]:
# 2.- Compare_models(): Comparar múltiples modelos y los rankers

best_model = compare_models()

,Model,MAE,MSE,RMSE,R2,RMSLE,MAPE,TT (Sec)
lr,Linear Regression,11281.2714,215464568.7192,14435.3818,0.9593,0.0583,0.0449,0.3190
lasso,Lasso Regression,11280.1557,216043061.6317,14454.2273,0.9592,0.0583,0.0449,0.2080
lar,Least Angle Regression,11280.0105,216055877.4185,14454.6787,0.9592,0.0583,0.0449,0.0100
llar,Lasso Least Angle Regression,11280.1667,216041830.3268,14454.1746,0.9592,0.0583,0.0449,0.0100
ridge,Ridge Regression,11566.6268,216281347.5003,14482.7539,0.9587,0.0582,0.0463,0.0100
huber,Huber Regressor,11821.9848,221718852.3957,14657.2633,0.9578,0.0588,0.0479,0.0090
et,Extra Trees Regressor,16349.6766,418476041.8270,20059.3949,0.9139,0.0913,0.0711,0.0210
ada,AdaBoost Regressor,17103.9362,466089934.9115,21110.0115,0.9065,0.0959,0.0749,0.0170
gbr,Gradient Boosting Regressor,17468.2462,481907001.7354,21311.7623,0.9048,0.0930,0.0742,0.0140
rf,Random Forest Regressor,17471.2278,495610198.0678,21670.6051,0.9004,0.0958,0.0741,0.0240


In [7]:
lr = create_model('lr') # Es el mejor modelo 

,MAE,MSE,RMSE,R2,RMSLE,MAPE
Fold,,,,,,
0,8228.1193,128207335.1395,11322.8678,0.9802,0.0408,0.0314
1,12382.6057,231795614.2626,15224.8354,0.9742,0.0808,0.0624
2,11402.7735,178327899.8759,13353.9470,0.9274,0.0491,0.0403
3,14086.9562,375781739.9028,19385.0907,0.9333,0.0633,0.0466
4,9469.3045,146238973.8241,12092.9307,0.9816,0.0354,0.0299
5,10738.6190,222111756.4850,14903.4143,0.9663,0.0422,0.0336
6,15475.3250,330392051.7883,18176.6898,0.9675,0.1052,0.0765
7,13786.7347,249341965.8060,15790.5657,0.9295,0.0524,0.0450
8,8076.4682,122365620.4205,11061.8995,0.9793,0.0422,0.0318


In [8]:
# 4.- Tune_model(): Ajuste automático de hiperparámetros

tuned_lr = tune_model(lr)

,MAE,MSE,RMSE,R2,RMSLE,MAPE
Fold,,,,,,
0,8228.1193,128207335.1395,11322.8678,0.9802,0.0408,0.0314
1,12382.6057,231795614.2626,15224.8354,0.9742,0.0808,0.0624
2,11402.7735,178327899.8759,13353.9470,0.9274,0.0491,0.0403
3,14086.9562,375781739.9028,19385.0907,0.9333,0.0633,0.0466
4,9469.3045,146238973.8241,12092.9307,0.9816,0.0354,0.0299
5,10738.6190,222111756.4850,14903.4143,0.9663,0.0422,0.0336
6,15475.3250,330392051.7883,18176.6898,0.9675,0.1052,0.0765
7,13786.7347,249341965.8060,15790.5657,0.9295,0.0524,0.0450
8,8076.4682,122365620.4205,11061.8995,0.9793,0.0422,0.0318


Fitting 10 folds for each of 2 candidates, totalling 20 fits
Original model was better than the tuned model, hence it will be returned. NOTE: The display metrics are for the tuned model (not the original one).


In [9]:
evaluate_model(tuned_lr)

interactive(children=(ToggleButtons(description='Plot Type:', icons=('',), options=(('Pipeline Plot', 'pipelin…

In [12]:
preds = predict_model(tuned_lr)

,Model,MAE,MSE,RMSE,R2,RMSLE,MAPE
0,Linear Regression,12666.9295,222599218.6786,14919.7593,0.9743,0.0807,0.0576


In [11]:
preds.head()

,area_m2,num_habitaciones,num_banos,distrito,antiguedad_anos,vista_mar,precio,prediction_label
8,87,2,2,San Isidro,19,0,242334.515625,229980.233894
70,100,4,2,Miraflores,5,1,300367.437500,323149.386211
82,31,1,1,Surco,14,1,122736.742188,96058.098024
28,97,3,2,Surco,12,1,315150.500000,298969.196595
63,130,1,1,Surco,0,1,364217.593750,361290.005458


In [13]:
# 8.- Save_model/load_model(): Guardar y cargar modelo

save_model(tuned_lr,'modelo_precio_final')

Transformation Pipeline and Model Successfully Saved


(Pipeline(memory=Memory(location=None),
          steps=[('numerical_imputer',
                  TransformerWrapper(include=['area_m2', 'num_habitaciones',
                                              'num_banos', 'antiguedad_anos',
                                              'vista_mar'],
                                     transformer=SimpleImputer())),
                 ('categorical_imputer',
                  TransformerWrapper(include=['distrito'],
                                     transformer=SimpleImputer(strategy='most_frequent'))),
                 ('onehot_encoding',
                  TransformerWrapper(include=['distrito'],
                                     transformer=OneHotEncoder(cols=['distrito'],
                                                               handle_missing='return_nan',
                                                               use_cat_names=True))),
                 ('normalize', TransformerWrapper(transformer=StandardScaler())),
        

| Métrica   | Valor          | Interpretación                                                                                                                         |
| --------- | -------------- | -------------------------------------------------------------------------------------------------------------------------------------- |
| **MAE**   | 12,666.93      | En promedio, las predicciones se equivocan en aproximadamente **12,667 unidades** de la variable objetivo.                             |
| **MSE**   | 222,599,218.68 | Error cuadrático medio. Penaliza fuertemente los errores grandes. Se usa más para optimización que para interpretación directa.        |
| **RMSE**  | 14,919.76      | El error típico de predicción es de aproximadamente **14,920 unidades**. Está en las mismas unidades que la variable dependiente.      |
| **R²**    | 0.9743         | El modelo explica el **97.43% de la variabilidad** de los datos. Es un ajuste excelente.                                               |
| **RMSLE** | 0.0807         | El error relativo en escala logarítmica es bajo. Indica que el modelo mantiene bien las proporciones entre valores grandes y pequeños. |
| **MAPE**  | 0.0576         | Error porcentual medio absoluto de **5.76%**. En promedio, las predicciones se desvían un 5.76% del valor real.                        |
